pkg/syncer/shared/finalizer.go

```diff
@@ -36,15 +39,23 @@ const (
 	SyncerFinalizerNamePrefix = "workload.kcp.dev/syncer-"
 )
 
-func EnsureUpstreamFinalizerRemoved(ctx context.Context, gvr schema.GroupVersionResource, upstreamClient dynamic.ClusterInterface, upstreamNamespace, syncTargetName string, logicalClusterName logicalcluster.Name, resourceName string) error {
-	upstreamObj, err := upstreamClient.Cluster(logicalClusterName).Resource(gvr).Namespace(upstreamNamespace).Get(ctx, resourceName, metav1.GetOptions{})
+func EnsureUpstreamFinalizerRemoved(ctx context.Context, gvr schema.GroupVersionResource, informers dynamicinformer.DynamicSharedInformerFactory, upstreamClient dynamic.ClusterInterface, upstreamNamespace, syncTargetName string, logicalClusterName logicalcluster.Name, resourceName string) error {
+	upstreamObjFromLister, err := informers.ForResource(gvr).Lister().ByNamespace(upstreamNamespace).Get(clusters.ToClusterAwareKey(logicalClusterName, resourceName))
 	if err != nil && !apierrors.IsNotFound(err) {
 		return err
 	}
 	if apierrors.IsNotFound(err) {
 		return nil
 	}
 
+	upstreamObj, ok := upstreamObjFromLister.(*unstructured.Unstructured)
+	if !ok {
+		klog.Errorf("Resource %s|%s/%s expected to be *unstructured.Unstructured, got %T", logicalClusterName.String(), upstreamNamespace, resourceName, upstreamObjFromLister)
+		return nil
+	}
+
+	upstreamObj = upstreamObj.DeepCopy()
```

Comment:
do deepcopy further down. This call is super heavy.

Reply:
:+1:

Reply:
Done